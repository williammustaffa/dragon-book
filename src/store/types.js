const types = {
  // User actions
  USER_LOGIN: "USER_LOGIN",
  USER_LOGIN_SUCCESS: "USER_LOGIN_SUCCESS",
  USER_LOGIN_FAILURE: "USER_LOGIN_FAILURE",
  USER_LOGOUT: "USER_LOGOUT",

  // User actions
  USER_CHECK_SESSION: "USER_CHECK_SESSION",
  USER_CHECK_SESSION_SUCCESS: "USER_CHECK_SESSION_SUCCESS",
  USER_CHECK_SESSION_FAILURE: "USER_CHECK_SESSION_FAILURE",

  // Profile creation
  CREATE_PROFILE: "CREATE_PROFILE",
  CREATE_PROFILE_SUCCESS: "CREATE_PROFILE_SUCCESS",
  CREATE_PROFILE_FAILURE: "CREATE_PROFILE_FAILURE"
};

export default types;