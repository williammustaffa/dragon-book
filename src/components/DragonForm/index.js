export { default as DragonForm } from "./DragonForm";