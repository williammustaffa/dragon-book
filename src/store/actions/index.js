// Session
export { default as userLogin } from "./user/userLogin";
export { default as userLoginSuccess } from "./user/userLoginSuccess";
export { default as userLoginFailure } from "./user/userLoginFailure";
export { default as userLogout } from "./user/userLogout";
export { default as userCheckSession } from "./user/userCheckSession";
export { default as userCheckSessionSuccess } from "./user/userCheckSessionSuccess";
export { default as userCheckSessionFailure } from "./user/userCheckSessionFailure";

// Profile
export { default as createProfile } from "./profiles/createProfile";
export { default as createProfileFailure } from "./profiles/createProfileFailure";
