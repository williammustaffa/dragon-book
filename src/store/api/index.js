
import database from "store/api/database";
import * as R from "ramda";
import { v1 as uuidv1 } from 'uuid';
import request from "./request";

const AUTH_TOKEN_KEY = "auth.token";

class API {
  /**
   * Fake API error delay
   * @param {String} value 
   * @param {Int} time 
   */
  static retrieveDataDelayed(value, time = 500) {
    return new Promise(resolve => {
      setTimeout(() => resolve(value), time);
    });
  }

  /**
   * Fake API error delay
   * @param {String} value 
   * @param {Int} time 
   */
  static retrieveErrorDelayed(value, time = 500) {
    return new Promise((resolve, reject) => {
      setTimeout(() => reject({ message: value }), time);
    });
  }

  /**
   * Create token for user
   * @param {*} param
   */
  static tokenize({ id, email }) {
    return btoa(JSON.stringify({ id, email }));
  }

  /**
   * Get user info from token
   */
  static getTokenizedInfo() {
    const token = localStorage.getItem(AUTH_TOKEN_KEY);
  
    return token && JSON.parse(atob(token));
  }

  /**
   * Parse response object
   * @param {*} res 
   */
  static parseReponse(res) {
    return res.data;
  }

  /**
   * Create a user profile
   */
  createProfile = (payload) => {
    const errors = [];

    const existingUserEmail = database.queryAll("profiles", {
      query: { email: payload.email }
    });

    const existingUserUsername = database.queryAll("profiles", {
      query: { username: payload.username }
    });

    if (existingUserEmail.length) {
      errors.push("The email address provided is already in use.");
    }

    if (existingUserUsername.length) {
      errors.push("The username provided is already in use.");
    }

    if (errors.length) {
      return API.retrieveErrorDelayed(errors.join("\r\n"));
    }

    database.insert("profiles", {
      id: uuidv1(),
      email: payload.email,
      username: payload.username,
      firstName: payload.firstName,
      lastName: payload.lastName,
      description: payload.description,
      password: payload.password,
      registrationDate: Date.now(),
    });

    database.commit();

    // Create user and login
    return this.userLogin(payload);
  }

  /**
   * Check user session an then populate data if needed
   */
  userCheckSession = () => {
    const data = API.getTokenizedInfo();

    const results = database.queryAll("profiles", {
      query: { id: data && data.id },
    });

    if (!results.length) {
      return API.retrieveErrorDelayed("No session found.");
    }

    const user = R.clone(results[0]);

    // No need to delay on first load
    return user;
  }

  /**
   * Do user login by given crendentials
   * @param {object} credentials 
   */
  userLogin({ email, password }) {
    const results = database.queryAll("profiles", {
      query: { email, password },
    });

    if (!results.length) {
      return API.retrieveErrorDelayed("You have entered an invalid username or password");
    }

    const user = R.clone(results[0]);

    // Fake a token storage
    localStorage.setItem(AUTH_TOKEN_KEY, API.tokenize(user));

    return API.retrieveDataDelayed(user, 3000);
  }

  /**
   * Clear user session when he loogs out
   */
  userLogout = () => {
    // Clear token
    localStorage.removeItem(AUTH_TOKEN_KEY);

    return API.retrieveDataDelayed(null, 0);
  }

  /**
   * List dragons
   */
  listDragons = () => {
    return request
      .get("/dragon")
      .then(API.parseReponse);
  }

  /**
   * Get dragon details
   * @param {String} id
   */
  getDragon = (id) => {
    return request
      .get(`/dragon/${id}`)
      .then(API.parseReponse);
  }

  /**
   * Get dragon details
   * @param {Object} data
   */
  createDragon = (data) => {
    return request
      .post("/dragon", data)
      .then(API.parseReponse);
  }

  /**
   * Update dragon details
   * @param {Object} data
   */
  updateDragon = (data) => {
    return request
      .put(`/dragon/${data.id}`, data)
      .then(API.parseReponse);
  }

  /**
   * Update dragon details
   * @param {String} data
   */
  deleteDragon = (id) => {
    return request
      .put(`/dragon/${id}`)
      .then(API.parseReponse);
  }
}

// Create a cached connector
let connectorInstance = null;

function getAPIConnector() {
  if (!connectorInstance) {
    connectorInstance = new API();
  }

  return connectorInstance;
}


export { API, getAPIConnector };