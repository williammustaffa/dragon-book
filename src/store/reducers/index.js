import { combineReducers } from "redux";
import { connectRouter } from "connected-react-router";
import dragons from "./dragons";
import dragon from "./dragon";
import user from "./user";

const createRootReducer = (history) => combineReducers({
  router: connectRouter(history),
  dragons,
  dragon,
  user,
});

export default createRootReducer;