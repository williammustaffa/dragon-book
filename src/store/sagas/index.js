import { all } from "redux-saga/effects";

// User
import { watchUserLoginAsync } from "./user/userLogin";
import { watchUserLogoutAsync } from "./user/userLogout";
import { watchUserCheckSessionAsync } from "./user/userCheckSession";
import { watchCreateProfileAsync } from "./user/createProfile";

// Dragons
import { watchFetchDragons } from "./dragons/fetchDragons";
import { watchFetchDragon } from "./dragon/fetchDragon";

/**
 * Root saga provided to redux middleware
 */
function* root() {
  const watchers = [];

  // User watchers
  watchers.push(watchUserLoginAsync());
  watchers.push(watchUserLogoutAsync());
  watchers.push(watchUserCheckSessionAsync());
  watchers.push(watchCreateProfileAsync());

  // Dragons watchers
  watchers.push(watchFetchDragon());
  watchers.push(watchFetchDragons());

  yield all(watchers);
}

export default root;