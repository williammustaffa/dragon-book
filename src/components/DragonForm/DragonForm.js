import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { Form, Message } from "semantic-ui-react";
import { useForm } from "react-hook-form";
import Dragon from "store/models/Dragon";
import * as R from "ramda";
import { CustomDropdown } from "components/CustomDropdown";

function DragonForm({ dragon = new Dragon(), onSubmit, loading, errorMessage }) {
  const { register, handleSubmit, errors, setValue, triggerValidation, getValues } = useForm({
    defaultValues: {
      id: dragon.id || "new",
      name: dragon.title || "",
      imageUrl: dragon.imageUrl || "",
      type: dragon.type || "",
      history: dragon.history || "",
    }
  });

  // Component did mount
  useEffect(() => {
    function validateImageUrl(imageUrl = "") {
      return /(http(s?):)([/|.|\w|\s|-])*\.(?:jpg|jpeg|gif|png)/i.test(imageUrl);
    }

    register({ name: "id" }, { required: false });
    register({ name: "name" }, { required: true });
    register({ name: "imageUrl" }, { required: true, validate: validateImageUrl });
    register({ name: "type" }, { required: true });
    register({ name: "history" }, { required: true });
  }, [register]);

  function getError(fieldName) {
    return !!errors[fieldName];
  };

  async function updateFormField(e, { name, value }) {
    setValue(name, value);
    await triggerValidation({ name });
  }

  function onFormSubmit(data) {
    if (typeof onSubmit === "function") {
      onSubmit(data);
    }
  }

  // Create type options
  const types = R.values(Dragon.types).map(type => ({
    key: type.id,
    text: type.name,
    value: type.id,
    label: { color: type.color, empty: true, circular: true },
  }));

  return (
    <Form onSubmit={handleSubmit(onFormSubmit)} loading={loading}>
      <Form.Input
        fluid
        defaultValue={dragon.name}
        name="name"
        onChange={updateFormField}
        error={getError("name")}
        label="Dragon Name"
      />
      <Form.Input
        fluid
        name="imageUrl"
        defaultValue={dragon.imageUrl}
        onChange={updateFormField}
        error={getError("imageUrl")}
        label="Image URL"
      />
      <CustomDropdown
        fluid
        selection
        name="type"
        defaultValue={dragon.type}
        onChange={updateFormField}
        options={types}
        error={getError("type")}
        label="Dragon Type"
      />
      <Form.TextArea
        name="history"
        defaultValue={dragon.history}
        onChange={updateFormField}
        error={getError("history")}
        label="History"
      />
      {
        errorMessage &&
        <Message negative>
          <Message.Header>Error</Message.Header>
          <p>{errorMessage}</p>
        </Message>
      }
      <div className="form-footer">
        <Form.Button floated="right" color="red">Save and publish</Form.Button>
      </div>
    </Form>
  )
}

DragonForm.propTypes = {
  dragon: PropTypes.object,
  loading: PropTypes.bool,
  errorMessage: PropTypes.string,
  onSubmit: PropTypes.func,
};

export default DragonForm;