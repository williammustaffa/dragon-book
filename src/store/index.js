import { applyMiddleware, compose, createStore } from "redux";
import { routerMiddleware } from "connected-react-router";
import createRootReducer from "./reducers";
import sagas from "./sagas";
import history from "./history";
import createSagaMiddleware from "redux-saga"
import logger from "redux-logger";
import configureStore from "redux-mock-store";

/**
 * Generates a redux store properly
 * applying reducers, states and middlewares
 *
 * @param {*} preloadedState 
 */
function generateStore(preloadedState) {
  const sagaMiddleware = createSagaMiddleware();

  const store = createStore(
    createRootReducer(history),
    preloadedState,
    compose(
      applyMiddleware(
        routerMiddleware(history),
        sagaMiddleware,
        logger,
      ),
    ),
  );

  // Run redux saga
  sagaMiddleware.run(sagas);

  return store;
}

/**
 * Mock store and return a mocked state
 * @param {*} preloadedState 
 */
function mockStore(preloadedState) {
  const middlewares = [];
  const generateMockStore = configureStore(middlewares);

  return generateMockStore(preloadedState);
}

export {
  generateStore,
  mockStore,
  history,
};