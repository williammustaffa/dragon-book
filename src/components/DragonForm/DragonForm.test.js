import React from "react";
import { Provider } from 'react-redux';
import { mockStore } from 'store';
import DragonForm from "./DragonForm";
import { render } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"

test("Test DragonForm component", () => {
  // Mock store
  const store = mockStore({});

  const { asFragment } = render(
    <Provider store={store}>
      <DragonForm />
    </Provider>
  );

  // Run tests
  expect(asFragment()).toMatchSnapshot();
});