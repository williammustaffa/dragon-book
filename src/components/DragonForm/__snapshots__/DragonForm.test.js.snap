// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DragonForm component 1`] = `
<DocumentFragment>
  <form
    class="ui form"
  >
    <div
      class="field"
    >
      <label>
        Dragon Name
      </label>
      <div
        class="ui fluid input"
      >
        <input
          aria-invalid="true"
          name="name"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="field"
    >
      <label>
        Image URL
      </label>
      <div
        class="ui fluid input"
      >
        <input
          aria-invalid="true"
          name="imageUrl"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="custom-dropdown field"
    >
      <label>
        Dragon Type
      </label>
      <div
        aria-expanded="false"
        class="ui fluid selection dropdown"
        name="type"
        role="listbox"
        tabindex="0"
      >
        <div
          aria-atomic="true"
          aria-live="polite"
          class="text"
          role="alert"
        >
          Common
        </div>
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
        >
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <div
              class="ui yellow circular empty label"
            />
            <span
              class="text"
            >
              Legendary
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <div
              class="ui purple circular empty label"
            />
            <span
              class="text"
            >
              Epic
            </span>
          </div>
          <div
            aria-checked="false"
            aria-selected="false"
            class="item"
            role="option"
            style="pointer-events: all;"
          >
            <div
              class="ui blue circular empty label"
            />
            <span
              class="text"
            >
              Rare
            </span>
          </div>
          <div
            aria-checked="true"
            aria-selected="true"
            class="active selected item"
            role="option"
            style="pointer-events: all;"
          >
            <div
              class="ui black circular empty label"
            />
            <span
              class="text"
            >
              Common
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="field"
    >
      <label>
        History
      </label>
      <textarea
        aria-invalid="true"
        name="history"
        rows="3"
      />
    </div>
    <div
      class="form-footer"
    >
      <div
        class="field"
      >
        <button
          class="ui red right floated button"
        >
          Save and publish
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
